README
WWW-NicoSound-Download version 1.11
===================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  - Readonly
  - Class::Accessor
  - HTML::DOM
  - Test::Exception

COPYRIGHT AND LICENCE

Copyright (C) 2010 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension WWW::NicoSound::Download.

1.12  Jul Sat  6 20:54:00 2011
    - STOP this module.
1.11  Sun Dec  5 19:43:12 2010
    - Needs Class::Accessor for installing this.
1.10  Sat Dec  4  1:32:20 2010
    - Needs Readonly for installing this.
1.09  Sun Nov 21 16:28:24 2010
    - Change API URL.
    - Change codes to simple.
    - Remove Error object -- i can't maintenace it.
1.08  Sat Mar 13 21:31:46 2010
    - sm12345678 is valid ID too.
1.07  Sat Mar  6 21:20:18 2010
    - so(\d{7}) is valid ID.
1.06  Thu Feb  4 19:09:22 2010
    - Specified the version of Test::More. It needs 0.94.
1.05  Fri Jan 29 11:57:00 2010
    - Supports Exception::Class. Replaced die to throw.
1.04  Tue Nov  3 14:36:13 2009
    - NicoSound's has ID. "zb0000000" is valid ID too.
1.03  Mon Nov  1  0:31:20 2009
    - Corresponded to the upgrade of HTML::DOM
      In ver.0.026, getElementById worked, but in ver.0.034 does not.
      Possibly, parsed HTML is not valid.  Change getElementById to getElementsByTagName.
1.02  Wed Oct 28 17:55:30 2009
    - Newly add the function named [can_find_homepage].
    - Do not test some functions
      when the environment can not reach to the NicoSound in 180(default) seconds.
1.01  Wed Oct 21 19:39:36 2009
    - If fail in the get_raw and the save_mp3 then die.
      In early version, those returns undef with warning message.
    - Fixed a bug.
      If ID was passed to "save_mp3", and The ID has be deleted,
      then the function preserves odd filename.
    - The failure when the NicoSound server is overworking was able to be detected.
    - Add tests for private methods.
0.05  Tue Sep 29 23:50:05 2009
    - Rewrite tests.  Split testfile by function.
    - Functions save_mp3, and get_raw can accept url that includes ID.
0.04  Mon Sep 14 14:48:12 2009
    - Add PREREQ_PM key in Makefile.PL.  That is HTML::DOM and Carp::Assert.
0.03  Mon Sep  7  9:36:58 2009
    - The get_id function returns ID if passed ID.
0.02  Mon Aug 24 19:58:24 2009
    - Removed limitation of Perl version.
0.01  Sat Aug 15 17:38:25 2009
    - original version; created by h2xs 1.23 with options
        -XAn WWW::NicoSound::Download



lib/WWW/NicoSound/Download/Error.pm
package WWW::NicoSound::Download::Error;

use utf8;
use strict;
use warnings;
use Exception::Class (
    # The root error class.
    "NicoSoundError" => {
        fields => [ qw( message ) ],
    },

    # Base error class for save_mp3, and get_raw.
    "E::CantSave" => {
        isa    => "NicoSoundError",
        fields => [ qw( message ) ],
    },

    "E::CantFindHomepage" => {
        isa    => "NicoSoundError",
        fields => [ qw( message status_line ) ],
    },

    "E::IDRequired" => {
        isa    => "NicoSoundError",
        fields => [ qw( message ) ],
    },

    "E::URLRequired" => {
        isa    => "NicoSoundError",
        fields => [ qw( message ) ],
    },

    "E::FilenameRequired" => {
        isa    => "NicoSoundError",
        fields => [ qw( message ) ],
    },

    "E::InvalidID" => {
        isa    => "NicoSoundError",
        fields => [ qw( message id ) ],
    },

    "E::InvalidURL" => {
        isa    => "NicoSoundError",
        fields => [ qw( message url ) ],
    },

    "E::InvalidFilename" => {
        isa    => "NicoSoundError",
        fields => [ qw( message filename ) ],
    },

    "E::FileDoesNotExist" => {
        isa    => "NicoSoundError",
        fields => [ qw( message filename ) ],
    },

    "E::DownloadError" => {
        isa    => "E::CantSave",
        fields => [ qw( message status_line ) ],
    },

    "E::HasBeenDeleted" => {
        isa    => "E::CantSave",
        fields => [ qw( message id ) ],
    },

    "E::Overworking" => {
        isa    => "E::CantSave",
        fields => [ qw( message id ) ],
    },

    "E::FileAlreadyExists" => {
        isa    => "E::CantSave",
        fields => [ qw( message filename ) ],
    },
);


1;
__END__
=head1 NAME

WWW::NicoSound::Download::Error - Error class for WWW::NicoSound::Download

=head1 SYNOPSIS

  use WWW::NicoSound::Download::Error;

  sub will_die {
      NicoSoundError->throw( message => "Somethins is wrong." );
  }

  eval { will_die( ) };

  if ( my $e = Exception::Class->caught( "NicoSoundError" ) ) {
      die $e;
  }

=head1 DESCRIPTION

Blah blah blah.

=head2 EXPORT

None by default.

=head1 DIAGNOSTICES

=over

=item NicoSoundError

This is top level class of error. This is root.

This has message field.

=item E::CantSave

This is base error class for save_mp3, and get_raw methods.

This has message field.

=item E::CantFindHomepage

This will be thrown when the environment can not reach to the NicoSound server.

This has message, and status_line fields.

=item E::IDRequired

This will be thrown when a ID parameter required.

This has message field.

=item E::URLRequired

This will be thrown when a URL parameter required.

This has message field.

=item E::FilenameRequired

This will be thrown when a filename parameter required.

This has message field.

=item E::InvalidID

This will be thrown when can not parse ID from URL.

This has message, and id fields.

=item E::InvalidURL

Same as above. This will be thrown when can not parse ID from URL.
This odd errors is because save_mp3 can accept URL.

This has message, and url fields.

=item E::InvalidFilename

This will be thrown when the filename is 0, or "".

This has message, and filename fields.

=item E::FileDoesNotExist

This will be thrown when the specified filename does not exist.

This has message, and filename fields.

=item E::DownloadError

This will be thrown when failed HTTP method of GET, or any.

This has message, and status_line fields.

=item E::HasBeenDeleted

This will be thrown when the ID's file has been deleted by NicoSound user.
Deleted file can not save.

This has message, and id fields.

=item E::Overworking

This will be thrown when the NicoSound server is overworking.
Try later. This often occurs at midnight.

This has message, and id fields.

=item E::FileAlreadyExists

This will be thrown when the file already exists.

This has message, and filename fields.

=back

=head1 SEE ALSO

WWW::NicoSound::Download

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.



lib/WWW/NicoSound/Download/Sound.pm
package WWW::NicoSound::Download::Sound;

use utf8;
use strict;
use warnings;
use Carp qw( croak );
use Readonly;
use base "Class::Accessor";
use File::Spec::Functions qw( splitdir );

Readonly my $INFO_URL  => "http://nicosound.anyap.info/sound/";
#Readonly my $SOUND_URL => "http://nicosound2.anyap.info:8080/sound/";
Readonly my $SOUND_URL => "http://ns1.anyap.info:8080/sound/";
Readonly my $ID_LIKE   => "([ns]m|zb|so) \\d{6,8}";

__PACKAGE__->mk_accessors(
    qw( url  id  title  filename ),
);

#
# Class methods.
#

sub new {
    my $class = shift;
    my %param = @_;
    my $self  = bless { }, $class;

    if ( exists $param{id} ) {
        die sprintf "Invalid ID[%s].", $param{id}
            unless $class->is_id_valid( $param{id} );

        $self->id( $param{id} );
    }
    elsif ( exists $param{url} ) {
        $self->url( $param{url} );
        $self->parse_id_from_url;
    }

    if ( exists $param{filename} ) {
        $self->filename( $param{filename} );
    }

    return $self;
}

sub is_id_valid {
    my $class = shift;
    my $id    = shift;

    die "ID required.",
        unless defined $id;

    return $id
        if $id =~ m{\A $ID_LIKE \z}msx;

    die sprintf "ID[%s] is invalid.", $id;
}

#
# Instance methods.
#

sub info_url {
    my $self = shift;

    return $INFO_URL . $self->id;
}

sub sound_url {
    my $self = shift;

    return $SOUND_URL . $self->id . ".mp3";
}

sub parse_id_from_url {
    my $self = shift;

    croak "URL required."
        unless defined $self->url;

    my $id = do {
        if ( $self->url =~ m{ ($ID_LIKE) (?:[^\d] | \z) }msx ) {
            my $candidate = $1;

            __PACKAGE__->is_id_valid( $candidate );

            $candidate;
        }
        else {
            die sprintf "URL[%s] is invalid.", $self->url;
        }
    };

    $self->id( $id );

    return $self;
}

sub parse_filename_from_title {
    my $self = shift;

    croak "Title required."
        unless $self->title;

    my $filename = join "_", splitdir( $self->title );

    if ( $filename !~ m{ [.] mp3 \z}imsx ) {
        $filename .= ".mp3";
    }

    $self->filename( $filename );

    return $self;
}

1;

__END__
=encoding utf-8

=head1 NAME

WWW::NicoSound::Download::Sound - A property class of MP3 file.

=head1 SYNOPSIS

  my $sound = $downloader->save( url => $url );
  say $sound->filename;

=head1 DESCRIPTION

This class keeps information of result of WWW::NicoSound::Download's downloader.

=head1 CLASS METHODS

=over

=item new

Returns new instance of this class.

=item is_id_valid

Checks the parameter likes nicosound ID.

=back

=head1 INSTANCE METHODS

=over

=item info_url

=item sound_url

=item parse_id_from_url

=item parse_filename_from_title

=back

=head1 PROPERTIES

=item url

=item id

=item title

=item filename

=over

=item url

=back

=head1 SEE ALSO

The homepage URL of NicoSound is "http://nicosound.anyap.info/"

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/WWW/NicoSound/Download.pm
package WWW::NicoSound::Download;

use utf8;
use strict;
use warnings;
use Carp qw( croak );
use Readonly;
use base "Class::Accessor";
use LWP::UserAgent;
use HTTP::Cookies;
use HTML::DOM;
use WWW::NicoSound::Download::Sound;

use constant IS_RIOT => 0;
use constant TIMEOUT => 10;
use constant DEBUG   => 0;

our $VERSION = "1.12";

Readonly my $MIN_BYTES => 1_000; # MP3 file should larger than this.
Readonly my %DEFAULT   => (
    ua      => LWP::UserAgent->new(
        timeout => TIMEOUT,
    ),
    is_riot => IS_RIOT,
    debug   => DEBUG,
);

#
# Class methods.
#

__PACKAGE__->mk_accessors(
    qw( ua  jar  is_riot  debug  sound ),
);

sub new {
    my $class = shift;
    my $self  = bless { }, $class;

    foreach my $name ( qw( ua is_riot debug ) ) {
        $self->_set_default( $name );
    }

    return $self;
}

sub url_to_id {
    my $class = shift;
    my $url   = shift
        or croak "URL required.";

    return eval { WWW::NicoSound::Download::Sound->new( url => $url )->id };
}

#
# Instance methods.
#

sub _set_default {
    my $self     = shift;
    my $property = shift
        or croak "Property required.";

    unless ( defined $self->$property ) {
        $self->$property( $DEFAULT{ $property } );
    }

    return $self;
}

sub save {
    my $self     = shift; 
    my %param    = @_;
    my $id       = $param{id};
    my $url      = $param{url};
    my $filename = $param{filename};

    my $sound = do {
        my %new;

        foreach my $key ( qw( id url filename ) ) {
            $new{ $key } = $param{ $key }
                if exists $param{ $key };
        }
        WWW::NicoSound::Download::Sound->new( %new );
    };

    $self->prepare( $sound );

    warn sprintf "--- Starting save_mp3[%s].\n", $sound->filename
        if $self->is_riot;

    my $res = $self->ua->get(
        $sound->sound_url,
        ":content_file" => $sound->filename,
    );

    if ( $res->is_error ) {
        die sprintf "Could not get MP3 data.[%d]", $res->status_line;
    }

    unless ( -f $sound->filename ) {
        die sprintf "Could not save MP3 data.[%s]", $sound->id;
    }

    # I hope the NicoSound server returns failure-code of HTTP-status.
    # Now, the server does not return failure code,
    # but returns HTML that the failure withers.
    if ( -T $sound->filename || -s $sound->filename < $MIN_BYTES ) {
=for comment
        unlink $sound->filename
            or die sprintf "The NicoSound server is overworking.\n"
                           . "Because of this, the downloaded file is not MP3 data.\n"
                           . "So I tried deleting, but it has failed.\n"
                           . "Please delete the file[%s].", $self->filename;
=cut

        die sprintf "The NicoSound server is overworking.\n"
                    . "### Try downloading later.[%s]", $sound->id;
    }

    return $sound;
}

sub prepare {
    my $self  = shift;
    my $sound = shift;

    croak "Sound required."
        unless $sound;

    # Pre-GET for save MP3.
    my $res = $self->ua->get( $sound->info_url );

    if ( $res->is_error ) {
        die sprintf "Failed pre GET that needs for get MP3.[%s]", $res->status_line;
    }

    warn "--- Succeeded first GET.\n"
        if $self->is_riot;

    # Can not get MP3 cause NicoSound web service.
    if ( $res->content =~ m{ video_deleted [.] jpg }imsx ) {
        die sprintf "The ID has been deleted.[%s]", $sound->id;
    }

    # Parse HTML.
    my $dom = HTML::DOM->new;
    $dom->write( $res->decoded_content );
    $dom->close;

    # Get title for filename.
    my $title = $dom->getElementsByTagName( "title" )->[0]->text;
    $title =~ s{\A \s* }{}msx;
    $title =~ s{ \s* \z}{}msx;
    my $hosi = "☆";
    my $nico = "にこ" . $hosi . "さうんど＃";
    $title =~ s{ [ ][-][ ]$nico .* }{}msx;
    $title =~ s{ [/] }{／}gmsx;

    # Did get title?
    if ( $title ) {
        $sound->title( $title );
    }
    else {
        die sprintf "Could not parse title from HTML.[%s]", $sound->id;
    }

    unless ( defined $sound->filename ) {
        $sound->parse_filename_from_title;
    }

    if ( -e $sound->filename ) {
        die sprintf "The file already exists.[%s]", $sound->filename;
    }

    warn sprintf "--- Title is [%s].\n", $sound->title
        if $self->is_riot;

    # The most important codes for NicoSound web service.
    # Cookies require for obtaining from NicoSound web service.
    # The flow of obtaining is:
    #   - GET  http://nicosound.anyap.info/sound/sm0000000
    #   - POST http://nicosound.anyap.info/sound/sm0000000
    #   - GET  http://nicosound2.anyap.info:8080/sound/sm0000000.mp3

    my $element_id    = "ctl00_ContentPlaceHolder1_SoundInfo1_btnLocal2";
    my( $post_param ) = grep { $_->id eq $element_id }
                        $dom->getElementsByTagName( "a" );

    # Does defined POST parameter?
    unless ( defined $post_param ) {
        die sprintf "Could not parse the parameters for POST.[%s]", $sound->id;
    }

    my( $event_target, $event_argument )
        = $post_param->href =~ m{ __doPostBack [(] ['] ([^']+) ['] [,] ['] ([^']*) }msx;

    # Does defined event target?
    unless ( $event_target ) {
        die sprintf "Could not parse the parameters for POST.[%s] - [%s]",
            $sound->id, $event_target;
    }

    my $cookie_jar = HTTP::Cookies->new;

    $res = $self->ua->post(
        $sound->info_url,
        {
            __EVENTTARGET   => $event_target,
            __EVENTARGUMENT => $event_argument,
        },
    );

    die "Status line isnt 302."
        if $self->debug && $res->status_line !~ m{302};

    warn "--- Succeeded the POST.\n"
        if $self->is_riot;

    $cookie_jar->extract_cookies( $res );

    if ( $cookie_jar ) {
        $self->ua->cookie_jar( $cookie_jar );
    }
    else {
        die "No jar."
            if $self->debug;
    }

    return $self;
}

1;

__END__
=encoding utf-8

=head1 NAME

WWW::NicoSound::Download - STOPPED

=head1 SYNOPSIS

  *STOPPED* don't use this.  Please wait next version, or please
  get this module from me.

  use WWW::NicoSound::Download;

  my $url = "http://nicosound.anyap.info/sound/smXXXXXXXX";

  my $downloader = WWW::NicoSound::Download->new;

  my $sound = $downloader->save( url => $url );

  print $sound->filename, "\n";

  use Path::Class qw( file );
  use HTML::SimpleLinkExtor;

  my $extor = HTML::SimpleLinkExtorj->new;
  my $html  = do {
      my $raw = file( "result_page_of_searching_from_nicosound.html" )->slurp;
      utf8::decode( $raw );
      $raw;
  };
  my @urls = do {
      $extor->parse( $raw );
      $extor->links;
  };
  my @ids = map { WWW::NicoSound::Download->url_to_id( $_ ) } @urls;

=head1 DESCRIPTION

This class is for downloading MP3 files from nicosound web service.

=head1 CLASS METHODS

=over

=item new

Returns a new instance of Downloader.

=item url_to_id

Parses and returns nicosound ID from url.

=back

=head1 INSTANCE METHODS

=over

=item save

Downloads a MP3 from nicosound, and returns instance of WWW::NicoSound::Download::Sound.

=back

=head1 PROPERTIES

=over

=item ua

Specifies a LWP::UserAgent's instance.

=item is_riot

is specified, report message what i'm doing.

=item debug

Tells me work strictry.

=back

=head1 SEE ALSO

The homepage URL of NicoSound is "http://nicosound.anyap.info/"

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut

